csrank/objectranking/lambdamart.py

```diff
@@ -0,0 +1,370 @@
+
+        """
+        #prepare array like features and imaginary qids
+        xdim = X.shape[0]  # n_instances - qid
+        ydim = X.shape[1]  # n_objects - documents
+        zdim = X.shape[2]  # n_features
+
+        features_as_list = deque()
+        for i in range(0,xdim):
+            for j in range(0,ydim):
+                row_as_list=deque([i])
+                features = deque()
+                for k in range(0, zdim):
+                    row_as_list.append(X[i, j, k])
+                features_as_list.append(row_as_list)
+
+        #Convert rankings to relevance scores     
+        scores_docsize = Y.shape[1]
+        relscore_train = np.subtract(scores_docsize, Y)
+
+        #prepare array like relevance score values
+        xdim_scores = relscore_train.shape[0]
+        ydim_scores = relscore_train.shape[1]
+
+        scores_as_list = deque()
+        for x in range(0,xdim_scores):
+            for y in range(0,ydim_scores):
+                scores_as_list.append(relscore_train[x,y])
+        
+        #Check if both the dimensions are the same
+        assert(len(features_as_list)==len(scores_as_list))
+        
+        #convert to numpy and resize the arrays 
+        features = np.asarray(features_as_list)
+        scores_unflat = np.array(scores_as_list)
+        scores = np.reshape(scores_unflat,(len(scores_unflat),1))
+
+        #Concatenate the reshaped arrays and return as trainin data
+        train_data = np.concatenate((scores,features),axis=1)
+
+        return train_data
+
+    def _i_dcg(self,scores,k=None):
+        """
+            Returns the ideal DCG value for the given scores
+        """
+        scores = [score for score in sorted(scores)[::-1]]
+        if k == None:
+            return np.sum([(np.power(2, scores[i]) - 1) / np.log2(i + 2) for i in range(len(scores))])
+
+        elif k > 0:
+            return np.sum([(np.power(2, scores[i]) - 1) / np.log2(i + 2) for i in range(len(scores[:k]))])
+
+    def _group_queries(self, training_data, qid_index):
+        query_indexes = {}
+        index = 0
+        for record in training_data:
+            query_indexes.setdefault(record[qid_index], [])
+            query_indexes[record[qid_index]].append(index)
+            index += 1
+        return query_indexes
+
+    def fit(self, X, y, **kwargs):
+        """
+            Fit a LambdaMART algorithm to the provided X and y arrays where X contains the features and y being the relevance scores.
+
+            Parameters
+            ----------
+            X : numpy array
+                (n_instances, n_objects, n_features)
+                Feature vectors of the objects
+            Y : numpy array
+                (n_instances, n_objects)
+                Rankings of the given objects
+            **kwargs
+                Keyword arguments for the fit function
```

Comment:
`**kwargs` is not used in the function. Maybe pass it on to the `DecisionTreeRegressor` and document it here?